crates/re_arrow_store/src/store_gc.rs

```diff
@@ -109,116 +118,163 @@ impl DataStore {
-                    // NOTE: We're dropping the bucket itself in this case, rather than just its
-                    // contents.
-                    dropped_size_bytes = bucket_size_bytes;
-                    table.buckets.remove(&bucket_key);
-
-                    if bucket_key == TimeInt::MIN {
-                        if let Some((_, bucket)) = table.buckets.pop_first() {
-                            table.buckets.insert(TimeInt::MIN, bucket);
-                        }
-                    }
+                num_bytes_to_drop -= table.try_drop_row(row_id, time.as_i64()) as f64;
+            }
+        }
+
+        row_ids
+    }
+}
+
+impl IndexedTable {
+    /// Tries to drop the given `row_id` from the table, which is expected to be found at the
+    /// specified `time`.
+    ///
+    /// Returns how many bytes were actually dropped, or zero if the row wasn't found.
+    fn try_drop_row(&mut self, row_id: RowId, time: i64) -> u64 {
+        let table_has_more_than_one_bucket = self.buckets.len() > 1;
+
+        let (bucket_key, bucket) = self.find_bucket_mut(time.into());
+        let bucket_num_bytes = bucket.size_bytes();
+
+        let mut dropped_num_bytes = {
+            let inner = &mut *bucket.inner.write();
+            inner.try_drop_row(row_id, time)
+        };
+
+        // NOTE: We always need to keep at least one bucket alive, otherwise we have
+        // nowhere to write to.
+        if table_has_more_than_one_bucket && bucket.num_rows() == 0 {
+            // NOTE: We're dropping the bucket itself in this case, rather than just its
+            // contents.
+            debug_assert!(
+                dropped_num_bytes <= bucket_num_bytes,
+                "Bucket contained more bytes than it thought"
+            );
+            dropped_num_bytes = bucket_num_bytes;
+            self.buckets.remove(&bucket_key);
+
+            // NOTE: If this is the first bucket of the table that we've just removed, we need the
+            // next one to become responsible for `-∞`.
+            if bucket_key == TimeInt::MIN {
+                if let Some((_, bucket)) = self.buckets.pop_first() {
+                    self.buckets.insert(TimeInt::MIN, bucket);
                 }
+            }
+        }
 
-                drop_at_least_size_bytes -= dropped_size_bytes as f64;
-                table.buckets_size_bytes -= dropped_size_bytes;
-                table.buckets_num_rows -= dropped_rows;
+        self.buckets_size_bytes -= dropped_num_bytes;
+        self.buckets_num_rows -= (dropped_num_bytes > 0) as u64;
+
+        dropped_num_bytes
+    }
+}
+
+impl IndexedBucketInner {
+    /// Tries to drop the given `row_id` from the table, which is expected to be found at the
+    /// specified `time`.
+    ///
+    /// Returns how many bytes were actually dropped, or zero if the row wasn't found.
+    fn try_drop_row(&mut self, row_id: RowId, time: i64) -> u64 {
+        self.sort();
```

Comment:
# ⚠️ Quadratic Behavior Detected!

Given in-order insertion, we always remove the first row of the first bucket. We do this iteratively, and each removal requires a resort. This leads to quadratic time complexity, though bounded by the maximum rows per bucket.

I see no easy way out of this, expect to replace all the `SmallVec`s of `IndexedBucketInner` with something more like `VecDeque` and special-case removing from the front, or do to something similar manually (keep the `SmallVec`s, but add a `initial_skip_rows: u64` member to `IndexedBucketInner`).

Since the problem is bounded by the max number of rows per bucket it is _probably_ fine, but it should at least be documented, especially where we configure the max rows per bucket (`DataStoreConfig::indexed_bucket_num_rows`), where we need to ad a big fat warning ("The GC costs increases quadratically with this number, so do not make it too large!!!"). Perhaps there should also be a run-time warning when we detect large buckets.

It would be useful to create a benchmark that hits this behavior using different values for `indexed_bucket_num_rows`

Reply:
Yes, that's what was alluded to by "Focused on correctness, millions of opportunities to improve performance in the future" in the intro.

There are plenty of workload specific optimizations possible, the `sort` itself hasn't been optimized at all, etc.

I'll add an issue.

> It would be useful to create a benchmark that hits this behavior using different values for `indexed_bucket_num_rows`

`datastore/num_rows=1000/num_instances=1000/packed=false/gc` should cover that already.